### ISSUE TITLE
New style of protocol API to match the limitation of Chromium's network stack

### ISSUE THREAD

**Issue by zcbenz** [atom/electron#2475]
Currently our protocol API allows returning different jobs for different requests with pure JavaScript, which is powerful but unfortunately above Chromium's ability. 

In Chromium the network request is started in IO thread and we have to respond to it in IO thread too, however JavaScript runs in UI thread. So when we received a network request, we have to hold on and switch to UI thread to ask user which job we should create, and then switch back to IO thread and handle the request.

The problem is this violates the design of Chromium's network stack, which requires the job to be created immediately when the request is made. We took a quite tricky way to work around this, and it has resulted in some bugs that can not be fixed (#1123).

Also after Chrome 44, our tricky way is crashing the program, so I have to change the design of the protocol module to match Chromium's limitation. The new design looks like this:

```javascript
protocol.registerFileProtocol('music', function(request, callback) {
  callback({path: '/tmp/music.mp3'});
}, function(error) {
  console.log('registered');
});

protocol.registerBufferProtocol('resource', function(request, callback) {
  callback({mimeType: 'text/html', data: new Buffer('text')});
}, function(error) {
  console.log('registered');
});
```

Unfortunately there is no way to be compatible with old APIs, since the original implementation is crashing and the new API is not so flex with original API. So this will be a breaking change.

**Comment by marcbachmann** [comment]
> The problem is this violates the design of Chromium's network stack, which requires the job to be created immediately when the request is made.

So the callback still needs to be called in the same tick?
That might be a bit confusing. But as long as it's documented, that's ok.

Do you like to remove the RequestJob classes completely? IMO that's a strange api anyways.

**Comment by zcbenz** [comment]
> So the callback still needs to be called in the same tick?

No, it is total asynchronous, the type of the request job (like buffer job or file job) has to been determined in the same tick, but the content of the request job (like the path of file job) doesn't have to.

> Do you like to remove the RequestJob classes completely?

Yeah.

**Comment by Tribex** [comment]
@zcbenz Is it possible to cancel a request using this API?

**Comment by deepak1556** [comment]
@Tribex yeah the new api is flexible to cancel any request job with error option in the handler [callback](https://github.com/atom/electron/commit/337460cdc278afbb2753680cf49b8442892db4bf).
```
callback({error: -300}) or callback(-300)
```

**Comment by Tribex** [comment]
@deepak1556 Excellent, thanks for clearing that up!

**Comment by paulcbetts** [comment]
Oof, this is a bummer for electron-compile but I think I can make most things work

**Comment by yan-foto** [comment]
Is there any way to use the default handler if one is intercepting a protocol? In the previous version you could just return `null` or `undefined` and it  would have taken care for you.

**Comment by deepak1556** [comment]
@yan-foto afraid that is not possible, in the previous version we fallback to create a requestjob from internal job factory of Electron if a handler was available for the corresponding scheme, this was possible because we created the requestjob only after handler was executed. In the new design it is required to determine the job early and once you have intercepted a scheme with `intercept<Http|File|String|Buffer>Protocol` you have decided the job for the response, hence it cannot be changed after handler was executed. 
It could be made possible with a new api `interceptDefaultProtocol` and have the request job created from default job factory handlers but in the end will be extending `JsAsker` into previous  `AdapterRequestJob` version, which is not good.

**Comment by yan-foto** [comment]
@deepak1556  I need some more time to get my head round this! I want to do [something pretty simple](https://github.com/yan-foto/electron-jade): compile all files with `.jade` extension and handle the rest as default. This works pretty find with non-binary files but fails on binary ones (e.g. `.png`). I use `interceptBufferProtocol` and return the content of the file as a `Buffer` with corresponding MIME type but it fails anyways :confounded:

**Comment by deepak1556** [comment]
@yan-foto need to create image buffer data from `native-image`

```
content = require('native-image').createFromPath(file).toPng()
```

**Comment by yan-foto** [comment]
@deepak1556 Sorry I was too impatient on my previous comment. At the moment [the following](https://github.com/yan-foto/electron-jade/blob/master/index.js) works fine:

```js
protocol.interceptBufferProtocol('file', function(request, callback) {
  var file = request.url.substr(7);
  var content = fs.readFileSync(file);
  var mime = require('mime');
  callback({data: content, mimeType: mime.lookup(ext)});
}, function () {});
```

I was reading the binary as text by passing `urtf8` as charset to `readFileSync`. When no charset is provided `readFileSync` returns exactly the `Buffer` I need.

I really appreciate your help.

**Comment by zcbenz** [comment]
Closed with #2474.
